### PR TITLE
Introduce SearchProcessor

### PR DESCRIPTION
**Pull request by sstok**
|Q            |A  |
|---          |---|
|Bug Fix?     |no |
|New Feature? |yes|
|BC Breaks?   |no |
|Deprecations?|no |
|Fixed Tickets|   |
|License      |MIT|

This replaces #1 

Todo:

- [x] Update/add translations
- [x] Tests
    - [x] SearchProcessor
    - [x] CacheSearchProcessor
    - [x] ~~ExceptionParsers~~
- [x] ~~Update documentation~~ (will do this in https://github.com/rollerworks/RollerworksSearch/issues/73 as it requires a fully functional error displayer).

### CODE REVIEW

**Comment by sstok** [comment]
The SearchFilterStorage (better name needed) keeps track of which search-code is used for which preference slot. Basically `slot-name => search-code`, and then storing this somewhere (eg. a session/cookie etc.) The filter-code is an exported search-condition.

The idea behind this that you have multiple searches on a single page, and restore them when needed. *When you return to a page and there is a search-code for the slot-name it will get the search-code and (optionally) redirect with the search-code applied or just applied directly.*

We are not using a serialized search-condition here because the search-code is user input, and any changes in the configuration should reprocess the original input.

**Comment by sstok** [comment]
I have a better idea for the Form, which fortunately is possible as bridge :+1: 

The Fields already support Views so its possible to use this for rendering form fields that will be used client-side. *Processing a big collection of Forms is a big risk for performance and may cause wrong results for transformers.* So having our own Form Templates and rendering will make this much easier :dancer:

File: src/Rollerworks/Bundle/SearchBundle/SearchFilterStorageInterface.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * SearchFilterStorageInterface for storing search-filters
+ * for persistent usage.
+ *
+ * A storage can be anything from a cookie, PHP session or
+ * entry in a database.
+ *
+ * @package Rollerworks\Bundle\SearchBundle
+ */
+interface SearchFilterStorageInterface
```
Comment (sstok) [inline src/Rollerworks/Bundle/SearchBundle/SearchFilterStorageInterface.php]: Rename to `SearchCodeStorage`

File: src/Rollerworks/Bundle/SearchBundle/Processor/SearchProcessorFactoryInterface.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * (c) 2014 Sebastiaan Stok <s.stok@rollerscapes.net>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Rollerworks\Bundle\SearchBundle\Processor;
+
+use Rollerworks\Bundle\SearchBundle\DependencyInjection\Factory\FieldSet;
```
Comment (sstok) [inline src/Rollerworks/Bundle/SearchBundle/Processor/SearchProcessorFactoryInterface.php]: This is wrong.

**Comment by cordoval** [comment]
what is introce?

**Comment by sstok** [comment]
The SearchCodeStorage will be done in another PR as this requires more work.
Its only about storing a search preference, this current PR has a higher priority.

**Comment by sstok** [comment]
When using the form its not possible to simply use `?filter` in the URI so make it possible to configure the form name as `my_form[some-sub][%s]` where %s replaced with the field-name we want (filter/format).

**Comment by sstok** [comment]
ValuesError path to translation, `'singleValues[0].value': 'Single value at index 0 "value"': Message`

File: src/Processor/AbstractSearchProcessor.php [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        $params = $query ? $this->request->query : $this->request->request;
+
+        if ($this->uriPrefix) {
+            $value = $params->get($this->uriPrefix.'['.$name.']', $default, true);
+        } else {
+            $value = $params->get($name, $default);
+        }
+
+        // Use default when invalid
+        if ($validator && !call_user_func($validator, $value)) {
```
Comment (sstok) [inline src/Processor/AbstractSearchProcessor.php]: This needs to be changed, to only is_string() for `$query === true` so we can also accept array input for an input processor.

Comment (sstok) [inline src/Processor/AbstractSearchProcessor.php]: Done

**Pull request merged**